doc/design/parallel-node-config.md

```diff
@@ -0,0 +1,249 @@
+---
+title: Parallel SR-IOV configuration
+authors:
+  - e0ne
+reviewers:
+  - adrianchiris
+  - SchSeba
+creation-date: 18-07-2023
+last-updated: 18-07-2023
+---
+
+# Parallel SR-IOV configuration
+
+## Summary
+Allow SR-IOV Network Operator to configure more than one node at the moment.
+
+## Motivation
+SR-IOV Network Operator configures SR-IOV one node at a time and one nic at a same time. That means we’ll need to wait
+hours or even days to configure all NICs  on large cluster deployments. With multi-NICs deployments Operator configures
+NICs one by one on each node which  leads to a lot of unnecessary node drain calls during SR-IOV configuration. Also
+moving all draining logic to a centralized place which will reduce chances of race conditions and bugs that were
+encountered before in sriov-network-config-daemon with draining.
+
+### Use Cases
+
+### Goals
+* Number of drainable nodes should be 1 be default
+* Number of drainable nodes should be configured by pool
+* Nodes pool should be defined by node selector
+* A node could be included into several pools. In this case pool should be 
+  selected by it’s priority like it’s implemented for SriovNetworkNodePolicy
+* Move all drain-related logic into the centralized place
+
+
+### Non-Goals
+Parallel NICs configuration on the same node is out of scope of this proposal
+
+## Proposal
+
+Introduce nodes pool drain configuration to meet following requirements:
+* Number of drainable nodes should be 1 be default
+* Number of drainable nodes should be configured by pool
+* Nodes pool should be defined by node selector
+* A node could be included into several pools. In this case pool should be selected by its priority like it’s
+  implemented for `SriovNetworkNodePolicy`
+
+
+### Workflow Description
+A new Drain controller will be introduced to manage node drain and cordon procedures. That means we don't need to do
+drain and use `drain lock` in config daemon anymore. The overall drain process will be covered by the following states:
+
+```golang
+DrainIdle        = "Idle"
+rainRequired     = "Drain_Required"
+DrainMcpPaused   = "Draining_MCP_Paused"
+Draining         = "Draining"
+DrainingComplete = "Draining_Complete"
+```
+
+Drain controller will watch for node annotation, `SriovNetworkNodeState` and `SriovNetworkPoolConfig` changes:
+```golang
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=SriovNetworkPoolConfig,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=SriovNetworkNodeState,verbs=get;list;watch;update;patch
+```
+
+Once `SriovNetworkNodeState.DrainStatus` will be marked as `Drain_Required` by config daemon controller add will start
+draining procedure.  Reconcile loop will check if current `Draining` nodes count is greater or equal than
+`maxUnavailable` and will mark it as `Draining` to start draining procedure. If current Draining nodes count equals to 
```

Comment:
Where is "maxUnavailable" taken from? Is this from a PodDisruptionBudget definition? If not, it should be made clear that a new parameter is added (since "maxUnavailable" is a common term in kubernates)

Reply:
It's a new field of in `SriovNetworkPoolConfig`

Reply:
Is it the same as `maxParallelNodeConfiguration` ? I'm fine with both names, just trying to get if they refer to the same concept

Reply:
Changed to `MaxParallelNodeConfiguration`